### ISSUE TITLE
Poor usability of regexpExtract

### ISSUE THREAD

**Issue by tavplubix** [ClickHouse/ClickHouse#56393]
See https://play.clickhouse.com/play?user=play#c2VsZWN0IHJlZ2V4cEV4dHJhY3QoYm9keSwgJycpIGFzIGxpbmssICogZnJvbSBnaXRodWJfZXZlbnRz

```
select regexpExtract(body, '') as link, * from github_events

"exception": "Code: 353. DB::Exception: Index value 1 is out of range, should be in [0, 1). (INDEX_OF_POSITIONAL_ARGUMENT_IS_OUT_OF_RANGE) (version 23.10.1.1954 (official build))"
```

**Comment by tavplubix** [comment]
Hm, actually it's just poor usability of `regexpExtract`, bad error message, and wrong error code

cc: @taiyang-li

**Comment by taiyang-li** [comment]
@tavplubix  I'll improve it

**Comment by tavplubix** [comment]
@taiyang-li, it's still bad: https://fiddle.clickhouse.com/cbf131a0-1e9d-48ab-a31a-55e001e67d89

The problem is that the default value of the third argument is 1. So it fails if I don't specify 0 explicitly. It doesn't make any sense. 

And the exception message is still misleading: I did not specify any indexes in the SELECT query and in the `regexpExtract` function. Maybe I did not even know that it has an optional third argument with the wrong default value. So how am I supposed to find out what's wrong with the query?

**Comment by taiyang-li** [comment]
> The problem is that the default value of the third argument is 1. So it fails if I don't specify 0 explicitly. It doesn't make any sense.
The default value is 1 for compatiability with Spark and MySQL

**Comment by taiyang-li** [comment]
I think we should add more information about the default value in exception message. I'll fix it in another PR

**Comment by taiyang-li** [comment]
@tavplubix  
 - spark: https://docs.databricks.com/en/sql/language-manual/functions/regexp_extract.html 
 - mysql: https://dev.mysql.com/doc/refman/8.0/en/regexp.html#function_regexp-substr 

> Do queries in MySQL and Spark fail if I don't specify the third argument of REGEXP_SUBSTR (we don't have this alias btw) or REGEXP_EXTRACT? 

It depends. If third argument in spark regexp_extract or fourth argument in mysql regexp_substr not specified, the match index is by default 1. It will fail if 1 is out of the range of regexp pattern. 

For regexp pattern `''`, match index 1 is out-of-range. Exceptions will be throw in spark and mysql. 


``` 
0: jdbc:hive2://localhost:10000/> SELECT regexp_extract('abc', '');
Error: org.apache.hive.service.cli.HiveSQLException: Error running query: java.lang.IllegalArgumentException: Regex group count is 0, but the specified group index is 1
``` 

```
> select REGEXP_SUBSTR('abc', ''); 
ERROR 3685 (HY000) at line 1: Illegal argument to a regular expression.
```

**Comment by tavplubix** [comment]
> If third argument in spark regexp_extract or fourth argument in mysql regexp_substr not specified, the match index is by default 1. It will fail if 1 is out of the range of regexp pattern.

@taiyang-li looks like regexp pattern indexes in MySQL are 1-based, so:
 - It's already incompatible with 0-based indexes in ClickHouse
 - 1 as a default value makes perfect sense for 1-based indexes in MySQL, but it's stupid for 0-based indexes in ClickHouse

Example: `select REGEXP_SUBSTR('abc', 'a.*');` succeeds and returns `abc` in MySQL: https://www.db-fiddle.com/f/oaSuiwafL2m3mRbpWQYscw/0
But in ClickHouse it fails. 

> For regexp pattern '', match index 1 is out-of-range.

It fails because the regular expression is empty, not because the index is out of range.

**Comment by taiyang-li** [comment]
@tavplubix yes, current implementation in CH keeps compatiability with Spark instead of MySQL indeed. I privately think we should not switch to MySQL style to avoid breaking backwards compatiability. 

If you have any ideas about improving usability of regexpExtract without breaking CH backwards compatiability, pls let me know, I'm glad to do it.

**Comment by tavplubix** [comment]
> setting 1 as default regexp pattern indexes makes sense because regexpExtract wants to extract something from input string

It doesn't make sense because it doesn't work. And tbh, I don't care much about compatibility with Spark

**Comment by alexey-milovidov** [comment]
@groeneai, fix this issue.

**Comment by groeneai** [comment]
On it. Plan: make the two-argument form `regexpExtract(haystack, pattern)` return the whole match (index 0) when the pattern has no capturing group, instead of throwing `INDEX_OF_POSITIONAL_ARGUMENT_IS_OUT_OF_RANGE`. The default index stays 1 when the pattern has at least one capturing group, so every currently-working query is unchanged - this only turns the error into a useful result. PR with a test shortly.

**Comment by clickgapai** [comment]
_Found via ClickGap automated review. Please close or comment if this is incorrect or needs adjustment._

### TL;DR
**Poor usability of regexpExtract**

### Reproduction
```
#!/bin/bash
set -uo pipefail
BIN="${1:?usage: repro.sh <clickhouse_binary>}"
out=$(timeout 30 "$BIN" local --query "SELECT regexpExtract('foobar', '') AS link" 2>&1)
rc=$?
echo "OBSERVED (exit=$rc): $out"
if [ "$rc" -eq 124 ]; then echo "TIMEOUT"; exit 124; fi
echo "$out" | grep -q "INDEX_OF_POSITIONAL_ARGUMENT_IS_OUT_OF_RANGE" && exit 1 || exit 0
```

### Bisect
- Not a regression — bug appears to be pre-existing or unbounded range. No bisect data.

### Affects
- **Reproduces on**: `master, 26.5, 26.4, 26.3, 25.8`

### Suggested next step
Check whether this is intended behavior, or a long-standing gap in `General` that the repro now exposes.

**Severity:** `P2`

---
_ClickGap · Finding: `h_issue_56393`_

**Comment by groeneai** [comment]
Opened https://github.com/ClickHouse/ClickHouse/pull/106374 for this.

`regexpExtract(haystack, pattern)` now returns the whole match when the pattern has no capturing group, instead of throwing `INDEX_OF_POSITIONAL_ARGUMENT_IS_OUT_OF_RANGE`. An explicit out-of-range index still throws, and patterns with a capturing group keep the default group 1.

**Comment by clickgapai** [comment]
_Found via ClickGap automated review. Please close or comment if this is incorrect or needs adjustment._

### TL;DR
**Poor usability of regexpExtract**

**ClickGap verified:** reproduced it with the script below on `26.5`; tested 6 version(s) — 5 affected, 1 clean (see Affects); long-standing (not a recent regression); introducing PR not yet pinned.

### Reproduction
```
SELECT regexpExtract('hello world', '') AS link;
-- On 26.1–26.5: Code: 353 DB::Exception: Index value 1 is out of range, should be in [0, 1). (INDEX_OF_POSITIONAL_ARGUMENT_IS_OUT_OF_RANGE)
-- On master (fixed by #106374): returns '' (the whole match, default index 0 when the pattern has no capture group)
```
[▶ Run on ClickHouse Fiddle](https://fiddle.clickhouse.com/a0b20b9a-1bcf-49c9-8d78-5fa7e34840fb)

### Bisect
- Reproduces on every release line tested down to `v24.1.8.22` (the oldest with a downloadable binary for this repro). The introducing change predates `v24.1.8.22`, so it can't be pinned from release binaries — this is a **long-standing** issue, not a recent regression. A maintainer with full git history can `git blame` the relevant file/line to find the exact change.

### Affects
- ✓ **Fix landed on master** in PR #106374. Branches below still have the bug and need the backport.
- **Reproduces on**: `26.5, 26.4, 26.3, 26.2, 26.1`
- **Does NOT reproduce on**: `master`
- **Backport needed**: 26.5, 26.4, 26.3, 26.2, 26.1

### Suggested next step
Backport PR #106374 to the affected release branches listed above. Bisect on `comp-regular-function` already identified the fix; no further triage needed.

**Component:** `comp-regular-function`
**Severity:** `P2`

cc fix PR reviewer: @antonio2368

---
_ClickGap · Finding: `h_issue_56393`_




### Triage metadata
- Primary component: `comp-regular-function`
- Component owners (`comp-regular-function`): @alexey-milovidov @bharatnc
